File: internal/authx/google.go
package authx

// Google service-account plumbing shared by the googleworkspace and
// googlecloud integrations: the key JSON, the GCE metadata token and the
// reason field of a Google API error.

import (
	"context"
	"crypto/rsa"
	"encoding/json"
	"errors"
	"net/http"
	"regexp"
	"time"

	"github.com/roee-hersh/hallpass/internal/httpx"
	"github.com/roee-hersh/hallpass/internal/integration"
)

// GoogleServiceAccountKey is the part of a service-account key JSON that
// signing an assertion needs.
type GoogleServiceAccountKey struct {
	ClientEmail  string
	PrivateKeyID string
	// TokenURI is the key's token endpoint, usually
	// https://oauth2.googleapis.com/token; it may be empty.
	TokenURI string
	Key      *rsa.PrivateKey
}

var googleEmailRe = regexp.MustCompile(`^[A-Za-z0-9!#$%&'*+/=?^_{|}~.-]{1,64}@[A-Za-z0-9.-]{1,255}$`)

// ParseGoogleServiceAccountKey decodes a service-account key JSON. Every
// failure is a credential_rejected error whose text never quotes the key.
func ParseGoogleServiceAccountKey(raw string) (GoogleServiceAccountKey, error) {
	var k struct {
		ClientEmail  string `json:"client_email"`
		PrivateKey   string `json:"private_key"`
		PrivateKeyID string `json:"private_key_id"`
		TokenURI     string `json:"token_uri"`
	}
	if err := json.Unmarshal([]byte(raw), &k); err != nil {
		return GoogleServiceAccountKey{}, integration.Wrap(integration.CodeCredentialRejected, err, "credential is not a service-account key JSON")
	}
	if !googleEmailRe.MatchString(k.ClientEmail) || k.PrivateKey == "" {
		return GoogleServiceAccountKey{}, integration.Errorf(integration.CodeCredentialRejected, "the service-account key JSON lacks client_email or private_key")
	}
	key, err := ParseRSAPrivateKey([]byte(k.PrivateKey))
	if err != nil {
		return GoogleServiceAccountKey{}, integration.Wrap(integration.CodeCredentialRejected, err, "the service-account private_key is not a PEM RSA key")
	}
	return GoogleServiceAccountKey{ClientEmail: k.ClientEmail, PrivateKeyID: k.PrivateKeyID, TokenURI: k.TokenURI, Key: key}, nil
}

// GoogleMetadataToken reads the attached service account's access token
// from the GCE metadata server at metadataURL (http://metadata.google.internal
// in production). now is the clock the expiry is computed from.
func GoogleMetadataToken(ctx context.Context, c *httpx.Client, metadataURL string, now func() time.Time) (Token, error) {
	if now == nil {
		now = time.Now
	}
	var out struct {
		AccessToken string          `json:"access_token"`
		ExpiresIn   json.RawMessage `json:"expires_in"`
	}
	resp, err := c.Do(ctx, &httpx.Request{Method: http.MethodGet,
		Path:   metadataURL + "/computeMetadata/v1/instance/service-accounts/default/token",
		Header: http.Header{"Metadata-Flavor": {"Google"}}})
	if err != nil {
		return Token{}, integration.Wrap(integration.CodeCredentialRejected, err, "the metadata server gave no token; auth_mode keyless needs a GCE or GKE Workload Identity")
	}
	if err := resp.JSON(&out); err != nil || out.AccessToken == "" {
		return Token{}, integration.Errorf(integration.CodeCredentialRejected, "the metadata server returned no access_token")
	}
	t := Token{Value: out.AccessToken}
	if secs := parseExpiresIn(out.ExpiresIn); secs > 0 {
		t.Expiry = now().Add(time.Duration(secs) * time.Second)
	}
	return t, nil
}

var googleReasonRe = regexp.MustCompile(`"reason"\s*:\s*"([A-Za-z_]+)"`)

// GoogleErrorReason extracts the first errors[].reason or details[].reason
// from the body snippet of a Google API error, or "" when there is none.
// Only the reason token is returned; the message is never used.
func GoogleErrorReason(err error) string {
	var se *httpx.StatusError
	if !errors.As(err, &se) {
		return ""
	}
	return GoogleReasonIn(se.Snippet)
}

// GoogleReasonIn extracts the first reason token from a Google error body.
func GoogleReasonIn(body string) string {
	if m := googleReasonRe.FindStringSubmatch(body); m != nil {
		return m[1]
	}
	return ""
}

File: hallpass-py/src/hallpass/authx/google.py
"""Google service-account plumbing shared by the googleworkspace and
googlecloud integrations: the key JSON, the GCE metadata token and the
reason field of a Google API error."""

from __future__ import annotations

import re
import time
from collections.abc import Callable
from dataclasses import dataclass
from typing import Any

from hallpass.authx.jwt import parse_rsa_private_key
from hallpass.authx.oauth2 import parse_expires_in
from hallpass.authx.token import Token
from hallpass.authx.util import RAW, STR, go_unmarshal
from hallpass.core.context import Context
from hallpass.core.decision import Code, errorf, wrap_error
from hallpass.core.errors import as_error
from hallpass.net import httpx

__all__ = ["GoogleServiceAccountKey", "google_error_reason", "google_metadata_token", "google_reason_in", "parse_google_service_account_key"]


@dataclass
class GoogleServiceAccountKey:
    client_email: str
    private_key_id: str
    # The key's token endpoint, usually https://oauth2.googleapis.com/token; may be empty.
    token_uri: str
    key: Any


_EMAIL_RE = re.compile(r"[A-Za-z0-9!#$%&'*+/=?^_{|}~.-]{1,64}@[A-Za-z0-9.-]{1,255}")


def parse_google_service_account_key(raw: str) -> GoogleServiceAccountKey:
    """Decode a service-account key JSON. Every failure is a
    credential_rejected error whose text never quotes the key."""
    k, err = go_unmarshal(raw, (("client_email", STR), ("private_key", STR), ("private_key_id", STR), ("token_uri", STR)))
    if err is not None:
        raise wrap_error(Code.CREDENTIAL_REJECTED, ValueError(err), "credential is not a service-account key JSON")
    if not _EMAIL_RE.fullmatch(k["client_email"]) or k["private_key"] == "":
        raise errorf(Code.CREDENTIAL_REJECTED, "the service-account key JSON lacks client_email or private_key")
    try:
        key = parse_rsa_private_key(k["private_key"].encode("utf-8"))
    except ValueError as e:
        raise wrap_error(Code.CREDENTIAL_REJECTED, e, "the service-account private_key is not a PEM RSA key") from e
    return GoogleServiceAccountKey(k["client_email"], k["private_key_id"], k["token_uri"], key)


def google_metadata_token(ctx: Context, c: httpx.Client, metadata_url: str, now: Callable[[], float] | None = None) -> Token:
    """The attached service account's access token from the GCE metadata
    server (http://metadata.google.internal in production)."""
    now = now or time.time
    try:
        resp = c.do(
            ctx,
            httpx.Request(
                method="GET",
                path=metadata_url + "/computeMetadata/v1/instance/service-accounts/default/token",
                header={"Metadata-Flavor": "Google"},
            ),
        )
    except Exception as e:
        raise wrap_error(Code.CREDENTIAL_REJECTED, e, "the metadata server gave no token; auth_mode keyless needs a GCE or GKE Workload Identity") from e
    out, err = go_unmarshal(resp.body, (("access_token", STR), ("expires_in", RAW)), stream=True)
    if err is not None or out["access_token"] == "":
        raise errorf(Code.CREDENTIAL_REJECTED, "the metadata server returned no access_token")
    secs = parse_expires_in(out["expires_in"])
    return Token(out["access_token"], now() + secs if secs > 0 else None)


# RE2's \s is ASCII [\t\n\f\r ]; Python's matches any Unicode space.
_REASON_RE = re.compile(r'"reason"[\t\n\f\r ]*:[\t\n\f\r ]*"([A-Za-z_]+)"')


def google_error_reason(err: BaseException | None) -> str:
    """The first errors[].reason or details[].reason in the body snippet of
    a Google API error, or "". Only the reason token; never the message."""
    se = as_error(err, httpx.StatusError)
    if se is None:
        return ""
    return google_reason_in(se.snippet)


def google_reason_in(body: str) -> str:
    m = _REASON_RE.search(body)
    return m.group(1) if m else ""
